pkg/skaffold/build/local/jib.go

```diff
@@ -18,16 +18,110 @@ package local
+	if err != nil {
+		return "", errors.Wrap(err, "maven build failed")
+	}
+	return skaffoldImage, nil
+}
+
+// generateMavenCommand generates the command-line to pass to maven for building a
+// project found in `workspace`.  The resulting image is added to the local docker daemon
+// and called `skaffoldImage`.
+func generateMavenCommand(_ /*workspace*/ string, skaffoldImage string, a *v1alpha3.JibMavenArtifact) ([]string, error) {
+	if a.Module != "" {
+		// TODO: multi-module
+		return nil, errors.New("Maven multi-modules not supported yet")
+	}
+	// use mostly-qualified plugin ID in case jib is not a configured plugin
+	commandLine := []string{"prepare-package", "com.google.cloud.tools:jib-maven-plugin::dockerBuild", "-Dimage=" + skaffoldImage}
+	if a.Profile != "" {
+		commandLine = append(commandLine, "-P"+a.Profile)
+	}
+	return commandLine, nil
+}
+
+func (b *Builder) buildJibGradle(ctx context.Context, out io.Writer, workspace string, a *v1alpha3.JibGradleArtifact) (string, error) {
+	skaffoldImage := generateJibImageRef(workspace, a.Project)
+	gradle, err := findBuilder("gradle", "gradlew", workspace)
+	if err != nil {
+		return "", errors.Wrap(err, "Unable to find gradle executable")
+	}
+	gradleCommand := generateGradleCommand(workspace, skaffoldImage, a)
+	commandLine := append(gradle, gradleCommand...)
+
+	err = executeBuildCommand(ctx, out, workspace, commandLine)
+	if err != nil {
+		return "", errors.Wrap(err, "gradle build failed")
+	}
+	return skaffoldImage, nil
+}
+
+// generateGradleCommand generates the command-line to pass to gradle for building an
+// project in `workspace`.  The resulting image is added to the local docker daemon
+// and called `skaffoldImage`.
+func generateGradleCommand(_ /*workspace*/ string, skaffoldImage string, a *v1alpha3.JibGradleArtifact) []string {
+	command := []string{}
+	if a.Project == "" {
+		command = append(command, ":jibDockerBuild")
+	} else {
+		// multi-module
+		command = append(command, ":"+a.Project+":jibDockerBuild")
+	}
+	command = append(command, "--image="+skaffoldImage)
+	return command
+}
+
+// executeBuildCommand executes the command-line with the working directory set to `workspace`.
+func executeBuildCommand(ctx context.Context, out io.Writer, workspace string, commandLine []string) error {
+	logrus.Infof("Building %v: %v", workspace, commandLine)
+	cmd := exec.CommandContext(ctx, commandLine[0], commandLine[1:]...)
+	cmd.Dir = workspace
+	cmd.Stdout = out
+	cmd.Stderr = out
+	return util.RunCmd(cmd)
 }
 
-func (b *Builder) buildJibGradle(_ /*ctx*/ context.Context, _ /*out*/ io.Writer, _ /*workspace*/ string, _ /*a*/ *v1alpha3.JibGradleArtifact) (string, error) {
-	return "", errors.New("buildJibGradle is unimplemented")
+// jibBuildImageRef generates a valid image name for the workspace and project.
+// The image name is always prefixed with `jib`.
+func generateJibImageRef(workspace string, project string) string {
+	imageName := "jib" + workspace
```

Comment:
what happens when `workspace == "."`?

Reply:
Because it's not a valid docker image name, it will be encoded as a `jib__d8c7cbe8892fe8442b7f6ef42026769ee6a01e67`.